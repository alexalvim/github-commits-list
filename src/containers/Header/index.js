import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Autocomplete from 'react-autocomplete';

import UserListItem from '../../components/UserListItem'
import logo from '../../imgs/light-logo.png';
import {
  searchUserRequest,
  getUserRepositoriesRequest,
  getUserRequest
} from '../../actions/user';
import {
  ContentWrapper,
  Container,
  Logo,
  autocompleteInputStyle,
  autocompleteWrapperStyle
} from './styles';

class Header extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      userField: '',
    }
  }

  handleChangeUserInput = (evt) => {
    const { searchUserRequest } = this.props;
    const user = evt.currentTarget.value;
    this.setState({
      userField: user
    });
    if(user.length > 0) {
      searchUserRequest(user);
    }
  }

  handleOnClickUserListItem = (login, avatarUrl, evt) => {
    const { getUserRepositoriesRequest, getUserRequest } = this.props;
    this.setState({
      userField: ''
    });
    getUserRepositoriesRequest(login, avatarUrl, 1);
    getUserRequest(login);
  }

  render() {
    const { searchedUsers } = this.props;
    const { userField } = this.state;
    const { handleChangeUserInput, handleOnClickUserListItem } = this;
    return (
      <ContentWrapper>
        <Container>
          <Logo src={logo} />
          <Autocomplete
            getItemValue={(user) => user.login}
            items={searchedUsers}
            renderItem={(user) =>
              <div key={user.id}>
                <UserListItem
                 title={user.login}
                 imgUrl={user.avatar_url}/>
              </div>
            }
            value={userField}
            onChange={handleChangeUserInput}
            inputProps={{...autocompleteInputStyle, placeholder: 'Digite o usuÃ¡rio a ser buscado'}}
            wrapperStyle={autocompleteWrapperStyle}
            onSelect={(_, user) => handleOnClickUserListItem(user.login, user.avatar_url)}
          />
        </Container>
      </ContentWrapper>
    );
  }
}

const mapStateToProps = store => ({
  searchedUsers: store.userState.searchedUsers,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ searchUserRequest, getUserRepositoriesRequest, getUserRequest }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Header);
