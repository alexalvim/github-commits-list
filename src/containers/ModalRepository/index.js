import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import InfiniteScroll from 'react-infinite-scroll-component';
import { IoMdSearch } from 'react-icons/io'

import {
  clearRepository,
  searchCommitRequest,
  getRepositoryCommitsRequest
} from '../../actions/repository';
import SimpleItem from '../../components/SimpleItem';
import ModalBox from '../../components/ModalBox';
import TextField from '../../components/TextField';
import TextButton from '../../components/TextButton';
import IconButton from '../../components/IconButton'
import {
  CommitsList,
  CommitsListWrapper,
  CommitsForm,
  commitTextFieldStyle,
  clearFilterButtonStyle
} from './styles';

class ModalRepository extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      commitSearchField: '',
      showClearFilter: false
    }
  }

  handleSubmitCommitSearch = (evt) => {
    evt.preventDefault();
    const { searchCommitRequest, user, repository } = this.props;
    const { commitSearchField } = this.state;
    if(commitSearchField.trim().length > 0) {
      searchCommitRequest(user.login, repository.name, commitSearchField);
      this.setState({
        showClearFilter: true
      });
    }
  }

  handleCloseModal = () => {
    const { closeModal, clearRepository } = this.props;
    this.setState({
      commitSearchField: '',
    })
    clearRepository();
    closeModal()
  }

  handleChangeCommitSearchField = (evt) => {
    const value = evt.currentTarget.value;
    this.setState({
      commitSearchField: value
    });
  }

  handleClearCommitFilters = () => {
    const { user, repository, getRepositoryCommitsRequest } = this.props;
    getRepositoryCommitsRequest(user.login, repository.name, 1)
    this.setState({
      showClearFilter: false,
      commitSearchField: ''
    });
  }

  handleSeeMoreCommits = () => {
    const { user, repository, getRepositoryCommitsRequest } = this.props;
    getRepositoryCommitsRequest(user.login, repository.name, repository.commitsPage + 1);
  }

  render() {
    const { repository, isOpen } = this.props
    const { commitSearchField, showClearFilter } = this.state;
    const {
      handleCloseModal,
      handleSubmitCommitSearch,
      handleChangeCommitSearchField,
      handleClearCommitFilters,
      handleSeeMoreCommits
    } = this;
    return (
      <ModalBox
        title={repository.name}
        isOpen={isOpen}
        closeModal={handleCloseModal}>
        <Fragment>
          <CommitsForm
            onSubmit={handleSubmitCommitSearch}>
            <TextField
              value={commitSearchField}
              placeholder='Digite aqui um termo para filtrar'
              onChange={handleChangeCommitSearchField}
              wrapperStyle={commitTextFieldStyle}/>
            <IconButton
              onClick={handleSubmitCommitSearch}
              Icon={IoMdSearch}/>
          </CommitsForm>
          {showClearFilter &&
            <TextButton
              wrapperStyle={clearFilterButtonStyle}
              onClick={handleClearCommitFilters}
              label='Limpar Filtros'/>}
          {repository.isLoadingCommits ?
            <span>Carregando Commits...</span>:
            repository.commits && repository.commits.length > 0 ?
              <CommitsListWrapper id='commits-list-wrapper'>
                <InfiniteScroll
                  scrollableTarget='commits-list-wrapper'
                  dataLength={repository.commits && repository.commits.length}
                  next={handleSeeMoreCommits}
                  hasMore={repository.hasNextPage}
                  loader={<span>Carregando Commits...</span>}>
                  <CommitsList>
                    {repository.commits.map((commit) => 
                      <li key={commit.sha}>
                      <SimpleItem
                        title={commit.commit.message}
                        description={commit.sha}/>
                      </li>)}
                  </CommitsList>
                </InfiniteScroll>
              </CommitsListWrapper>
            : <span>Commits n√£o encontrados</span>
          }
        </Fragment>
      </ModalBox>
    )
  }
}

const mapStateToProps = store => ({
  repository: store.repositoryState,
  user: store.userState,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({
    clearRepository,
    searchCommitRequest,
    getRepositoryCommitsRequest
  }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(ModalRepository);