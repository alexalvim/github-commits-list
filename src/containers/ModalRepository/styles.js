import styled from 'styled-components';

import Spaces from '../../spaces';

const CommitsList = styled.ul`
  padding-right: ${Spaces.base};

  > li {
    word-break: break-all;
  }
`;

const CommitsListWrapper= styled.div`
  margin-top: ${Spaces.base};
  overflow: auto;
`;

const CommitsForm = styled.form`
  display: flex;
  margin-bottom: ${Spaces.base};
`;

const commitTextFieldStyle = `
  margin-right: ${Spaces.half};
  width: 100%;
`;

const clearFilterButtonStyle = `
  align-self: center;
  margin-bottom: ${Spaces.base};
`;

export {
  CommitsList,
  CommitsListWrapper,
  CommitsForm,
  commitTextFieldStyle,
  clearFilterButtonStyle
}