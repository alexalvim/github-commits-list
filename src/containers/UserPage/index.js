import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import Header from '../Header';
import UserBox from '../../components/UserBox';
import ContentBox from '../../components/ContentBox';
import SimpleItem from '../../components/SimpleItem';
import { searchUserRequest, userRepositoriesRequest } from '../../actions/user';
import {
  Container,
  ContentHolder,
  RepositoriesList,
  contentBoxWrapperStyle
} from './styles';

class UserPage extends React.Component {

  handleClickSeeMore = () => {
    const { user, userRepositoriesRequest } = this.props;
    userRepositoriesRequest(user.login, user.avatarUrl, user.repositoriesPage + 1);
  }

  render() {
    const { user } = this.props;
    const { handleClickSeeMore } = this;
    console.log(user)
    return (
      <Fragment>
        <Header/>
        <Container>
          {user.login && user.login !== '' ?
          <ContentHolder>
            <UserBox
              avatarUrl={user.avatarUrl}
              login={user.login}/>
            <ContentBox
             title='Repositórios'
             wrapperStyles={contentBoxWrapperStyle}>
               {user.repositories &&
                user.repositories.length > 0 ?
                <Fragment>
                  <RepositoriesList>
                    {
                     user.repositories.map((repository) =>
                      <li key={repository.id}>
                        <SimpleItem
                          title={repository.name}
                          description={repository.description}/>
                      </li>
                    )}
                  </RepositoriesList>
                  {user.repositories.length < user.repositoriesCount &&
                    <span onClick={handleClickSeeMore}>ver mais</span>}
                </Fragment>
                : <p>Usuário não possui repositórios</p>}
            </ContentBox>
          </ContentHolder>
          : <p>Aplicação para verificar os repositórios e commits de usuários do Github. <br/>Use o campo acima para buscar um usuário</p>}
        </Container>
      </Fragment>
    );
  }
}

const mapStateToProps = store => ({
  user: store.userState,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ searchUserRequest, userRepositoriesRequest }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(UserPage);
