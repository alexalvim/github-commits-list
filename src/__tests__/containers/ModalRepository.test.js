import React from 'react';
import { shallow, mount } from 'enzyme';
import 'jest-styled-components'

import { ModalRepository } from '../../containers/ModalRepository';

const repository = {
  name: 'repositoryNameTest',
  commitsPage: 2
}

const user = {
  login: 'loginTest'
}

describe('Testing ModalRepository', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<ModalRepository user={user} repository={repository} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Testing ModalRepository handle functions', () => {
  it('should change the state value and call searchCommitRequest when call handleSubmitCommitSearch with value largest than zero', () => {
    const mockedEvt = {
      preventDefault: () => ({})
    };
    const mockFunction = jest.fn()
    const wrapper = shallow(<ModalRepository user={user} repository={repository} searchCommitRequest={mockFunction} />);

    wrapper.instance().state.showClearFilter = false;
    wrapper.instance().state.commitSearchField = 'test';
    wrapper.instance().handleSubmitCommitSearch(mockedEvt);
    expect(wrapper.instance().state.showClearFilter).toEqual(true);
    expect(mockFunction).toHaveBeenCalled();
  });

  it('shouldnt change the state value and dont call searchCommitRequest when call handleSubmitCommitSearch with value equal than zero', () => {
    const mockedEvt = {
      preventDefault: () => ({})
    };
    const mockFunction = jest.fn()
    const wrapper = shallow(<ModalRepository user={user} repository={repository} searchCommitRequest={mockFunction} />);

    wrapper.instance().state.showClearFilter = false;
    wrapper.instance().state.commitSearchField = '';
    wrapper.instance().handleSubmitCommitSearch(mockedEvt);
    expect(wrapper.instance().state.showClearFilter).toEqual(false);
    expect(mockFunction).toHaveBeenCalledTimes(0);
  });

  it('should change the state values and call clearRepository and closeModal when call handleCloseModal', () => {
    const mockFunction = jest.fn()
    const wrapper = shallow(
      <ModalRepository
        user={user}
        repository={repository}
        closeModal={mockFunction}
        clearRepository={mockFunction} />
    );

    wrapper.instance().state.showClearFilter = true;
    wrapper.instance().state.commitSearchField = 'test';
    wrapper.instance().handleCloseModal();
    expect(wrapper.instance().state.showClearFilter).toEqual(false);
    expect(wrapper.instance().state.commitSearchField).toEqual('');
    expect(mockFunction).toHaveBeenCalledTimes(2);
  });

  it('should change the state value when call handleChangeCommitSearchField', () => {
    const mockedEvt = {
      currentTarget: {
        value: 'foo'
      }
    };
    const wrapper = shallow(<ModalRepository user={user} repository={repository} />);

    wrapper.instance().state.commitSearchField = 'bar';
    wrapper.instance().handleChangeCommitSearchField(mockedEvt);
    expect(wrapper.instance().state.commitSearchField).toEqual('foo');
  });

  it('should change the state values and call getRepositoryCommitsRequest when call handleClearCommitFilters', () => {
    const mockFunction = jest.fn()
    const wrapper = shallow(
     <ModalRepository
      user={user}
      repository={repository}
      getRepositoryCommitsRequest={mockFunction} />
    );

    wrapper.instance().state.showClearFilter = true;
    wrapper.instance().state.commitSearchField = 'test';
    wrapper.instance().handleClearCommitFilters();
    expect(wrapper.instance().state.showClearFilter).toEqual(false);
    expect(wrapper.instance().state.commitSearchField).toEqual('');
    expect(mockFunction).toHaveBeenCalledWith('loginTest', 'repositoryNameTest', 1);
  });

  it('should call getRepositoryCommitsRequest when call handleSeeMoreCommits', () => {
    const mockFunction = jest.fn()
    const wrapper = shallow(
     <ModalRepository
      user={user}
      repository={repository}
      getRepositoryCommitsRequest={mockFunction} />
    );

    wrapper.instance().handleSeeMoreCommits();
    expect(mockFunction).toHaveBeenCalledWith('loginTest', 'repositoryNameTest', repository.commitsPage + 1);
  });
});