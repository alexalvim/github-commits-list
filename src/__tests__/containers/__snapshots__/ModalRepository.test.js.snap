// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ModalRepository should render correctly 1`] = `ShallowWrapper {}`;
